py/dml/codegen.py

```diff
@@ -943,6 +944,101 @@ def add(self, method):
 
 simple_events = SimpleEvents()
 
+def check_designated_initializers(site, etype, init_asts):
+    shallow_real_etype = safe_realtype_shallow(etype)
+    duplicates = set()
+    bad_fields = set()
+    remaining = set(shallow_real_etype.members)
+    for (field, init) in init_asts:
+        if field not in shallow_real_etype.members:
+            bad_fields.add(field)
+        elif field not in remaining:
+            duplicates.add(field)
+        else:
+            remaining.remove(field)
+    if (duplicates or bad_fields or remaining):
+        raise EDATAINIT(site,
+                        ''.join("\nduplicate initializer for "
+                                + f"member '{field}'"
+                                for field in duplicates)
+                        + ''.join(f"\n'{field}' is not a member of "
+                                  + f"{etype}"
+                                  for field in bad_fields)
+                        + ("\nmissing initializer(s) for field(s) "
+                           + ", ".join(f"'{field}'"
+                                       for field in remaining)
+                           if remaining else '')
+                        )
+
+def loss_on_truncation(v, bits, signed):
+    if not math.isfinite(v) or int(v) != v:
+        return True
+    # (-128).bit_length() == 8, which is not true for signed integers
+    # (bit_length() excludes sign bit).
+    # Adding one for signed negatives compensates for this
+    v = int(v) + (signed and v < 0)
+    # Don't indicate truncation loss on assignments of 1 to int1:s
+    if v == 1 and bits == 1 and signed:
+        return False
+    # Subtract one for signed numbers to compensate for the sign bit
+    return bits - signed < v.bit_length()
```

Comment:
Pondering if something like this would be more direct (after the uint1 check):
```
if signed:
    return (-1 << (bits - 1)) <= v < (1 << (bits - 1))
else:
    return (-1 << (bits - 1)) <= v < (1 << bits)
```
or maybe mask out the bits we lose and see if they contain information:
```
if signed:
    ...
else:
    trunc_mask = (-1 << bits)
    return (v & trunc_mask) == (trunc_mask if v < 0 else trunc_mask)
else:
```

Reply:
I could be ok with the first alternative. The second one I feel is worse than the others; less readable and understandable.

Reply:
It's stated in the comment above `loss_on_truncation`: to allow users to treat unsigned int types as the groups they are, `loss_on_truncation` also accepts any negation of representable values of the integer type. The max representable value for `uint8` is 255, so the smallest number accepted by `loss_on_truncation` is -255. -255 after truncation will become 1, which is indeed the inverse of 255. -256 is not accepted because 256 is not a representable value within `uint8`, so it doesn't make sense to talk about its inverse.

Reply:
That's a questionable argument. In situations when you think about an uint8 as a group in terms of arithmetic +, all negative numbers are just as much outside its range as 256 and would deserve a warning. (and note that if you use the Peano definition of 256, then it is defined as 255 + 1, which is valid also in the group ℤ<sub>256</sub>, making 256==0 as expected).

However, in cases where it is more natural to think of `uint8` as a group under the xor operator than under +, it is sometimes natural to assign a negative constant, because of the inherent signedness of small integers. Consider, for instance:
```
struct {uint2 x; uint2 y;} not_table = { {0, ~0}, {1, ~1}, {2, ~2}, {3, ~3} };
```
Here, your suggestion will warn for ~3, but not ~2.

(GCC is even more strict, `uint8 ok = ~127, warn = ~128`)

Reply:
After a brief discussion, `loss_on_truncation` now revamped to accept `~v` for any representable `v` when it comes to unsigneds.